cli: Add json output mode for 'check' and 'list' commands

detsch opened the pull request:
Implementation was straightforward, so I just made a first version while we get seek for some feedback

mike-sul commented on src/cli/cli.cc:
```diff
@@ -157,9 +160,59 @@ static CheckInResult checkIn(AkliteClientExt &client, CheckMode check_mode,
+      json_target["selected"] = true;
+      json_target["reason"] = gti_res.reason;
+    }
+    if (client.IsRollback(target)) {
```
This is not quite efficient performance wise since each `.IsRollback()` call does exactly the same action - get installed versions from the DB. I wonder if it is possible to optimize it somehow since a user device can have thousands targets?

detsch replied:
We could add a method to `LiteClient` returning a `set` with the names of all targets that caused a rollback. It would be just a  matter of using functionality from `get_known_but_not_installed_versions` and a little bit from `known_local_target`. What do you think?

mike-sul replied:
I thought about avoiding this iteration over targets here at all. Instead, we can extend the `filterTargets` functionality in the `AkliteClient::CheckIn()` to add the "bad" mark (another attribute of TufTarget object, or additional field in the custom json) (1).

The other two marks "current" and "selected" seems reasonable to set directly in the json output `getCheckInResultJson`. However, if we get more than one "client" for it, we may wanna add another two attributes to TufTarget in `AkliteClientExt::GetTargetToInstall`, but it means the input checkin result modification...(2)

I think it is more or less reasonable to do the first item now, unless it requires too much effort.

detsch replied:
I'm not sure about that, will give it a little more thought. Adding an attribute to `TufTarget`, or event to the custom json, would create an expectation on that attribute to always be correctly set, which has its price, while not being used that much.

mike-sul replied:
OK, let's not think about performance optimization at this stage.

detsch replied:
OK. This is something we can handle in a separate PR, as the potential performance hit is only related to the new feature.

mike-sul commented on src/cli/cli.cc:
```diff
@@ -157,9 +160,59 @@ static CheckInResult checkIn(AkliteClientExt &client, CheckMode check_mode,
+      json_target["reason"] = gti_res.reason;
+    }
+    if (client.IsRollback(target)) {
+      json_target["bad"] = true;
```
Maybe we should call it `rollback`, so it looks like `"rollback": true`, which is more clear way to indicate that a rollback occurred when trying to install a given target? Just IMO.

detsch replied:
I find the "rollback" term a bit ambiguous. I even changed some variable names because of that during the recent refactoring. "rollback" may cause confusion as there are 2 targets involved during a rollback operation. "caused_rollback" could be more appropriate to indicate the bad one. "bad" should be clear enough, but I'm open to change it.

mike-sul replied:
Maybe "failing" then, or just "failed".

detsch replied:
`failed` sounds good.

detsch replied:
Changed to `failed`

mike-sul commented on src/cli/cli.cc:
```diff
@@ -157,9 +160,59 @@ static CheckInResult checkIn(AkliteClientExt &client, CheckMode check_mode,
+    return res2StatusCode<CheckInResult::Status>(c2s, cr.status);
+  }
+
+  auto gti_res = client.GetTargetToInstall(cr);
```
I think we need to do this call only for the json output. If so, then it could be used only within "if true" branch.

mike-sul commented on src/cli/cli.cc:
```diff
@@ -184,11 +237,7 @@ StatusCode CheckIn(AkliteClientExt &client, const LocalUpdateSource *local_updat
-  }
 
-  auto gti_res = client.GetTargetToInstall(cr);
   if (gti_res && gti_res.selected_target.IsUnknown()) {
```
It looks like this is needed only for the json type of output?

detsch replied:
It is not related to the json output. This was added here: https://github.com/foundriesio/aktualizr-lite/pull/345/commits/487ef43eead80f51855691aae59705192431659f with the objective of allowing the CLI check command result to be used directly to decide if a pull/install is expected to be called. I.e., if there is a target that need to be installed.

mike-sul replied:
OK. Do we need it for the `list` too?

detsch replied:
I would keep it for the `list` operation as well. There is no change of state, and the adequate return code might be useful if someone is trying to use the currently stored metadata to decide on doing or not an update.

mike-sul replied:
ok.

mike-sul commented on src/main.cc:
```diff
@@ -67,22 +67,25 @@ static int status_main(LiteClient& client, const bpo::variables_map& unused) {
-  (void)unused;
-  return checkin(client, aklite::cli::CheckMode::Current);
+static int cli_list(LiteClient& client, const bpo::variables_map& params) {
+  return checkin(client, aklite::cli::CheckMode::Current, params);
```
I think we need to parameterize the `checkin` to indicate whether to do `GetTargetToInstall` or not because the `list` command should just list available targets. There is no need in `GetTargetToInstall` to determine whether a given target is "bad" or "current", only to determine whether it is selected or not. The latter is needed only for the `check` command.

detsch replied:
As discussed offline, I would rather keep calling `GetTargetToInstall` even for the `list` command, to leave the interface more consistent, and to allow the related functionality to be used even when not updating the TUF metadata.

detsch commented:
@mike-sul I've added two patches to better handle the log messages and `--json` option: Fixed the LOG_WARNING for non-root execution, and made loglevel "fatal" by default when using "--json 1".

mike-sul commented:
> @mike-sul I've added two patches to better handle the log messages and `--json` option: Fixed the LOG_WARNING for non-root execution, and made loglevel "fatal" by default when using "--json 1".

It works better now, thanks.
